File: leetcode/0276-paint-fence/solution.py
class Solution:
    def numWays(self, n: int, k: int) -> int:
        if n == 0:
            return 0
        if n == 1:
            return k

        same, diff = 0, k

        for _ in range(2, n + 1):
            new_same = diff
            new_diff = (same + diff) * (k - 1)
            same, diff = new_same, new_diff

        return same + diff

File: leetcode/0276-paint-fence/Solution.java
class Solution {
    public int numWays(int n, int k) {
        if (n == 0) return 0;
        if (n == 1) return k;

        long same = 0, diff = k;

        for (int i = 2; i <= n; i++) {
            long newSame = diff;
            long newDiff = (same + diff) * (k - 1);
            same = newSame;
            diff = newDiff;
        }

        return (int) (same + diff);
    }
}
